File: 0257_binary_tree_paths/binary_tree_paths.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


def binary_tree_paths(root: TreeNode) -> List[str]:
    """
    Time: O(n)
    Space: O(n)
    """
    res = []
    dfs(root, res, "")
    return res


def dfs(node: TreeNode, arr: List[str], curr: str):
    if not node:
        return True
    l = dfs(node.left, arr, curr + f"{node.val}->")
    r = dfs(node.right, arr, curr + f"{node.val}->")
    if l and r:
        arr.append(curr + str(node.val))

File: 0257_binary_tree_paths/binary_tree_paths.cpp
#include <iostream>
#include <string>
#include <vector>

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

vector<string> binaryTreePaths(TreeNode *root)
{
    vector<string> res;
    if (!root)
        return res;
    dfs(root, res, "");
    return res;
}

/**
 * Time: O(n)
 * Space: O(n) 
 */
void dfs(TreeNode *node, vector<string> &arr, string curr)
{
    if (!node->left && !node->right)
    {
        curr += to_string(node->val);
        arr.push_back(curr);
        return;
    }
    curr += to_string(node->val) + "->";
    if (node->left)
        dfs(node->left, arr, curr);
    if (node->right)
        dfs(node->right, arr, curr);
}
